functions/src/modules/gmail.functions.ts

```diff
@@ -0,0 +1,220 @@
+    oauth2Client.setCredentials(tokens);
+    await saveTokens(tokens);
+    res.send('Ok');
+  }
+);
+
+const fetchToken = async () => {
+  const queryResults = await db.ref(`/gmail/coordinator`).once('value');
+  const values = queryResults.val();
+
+  // https://github.com/GoogleCloudPlatform/cloud-functions-gmail-nodejs/blob/master/lib/oauth.js#L62
+  // https://github.com/googleapis/google-auth-library-nodejs/releases/tag/v2.0.0
+  if (
+    !values.oauth.expiry_date ||
+    // moment().valueOf() = Unix timestamp in milliseconds + 1 minute
+    values.oauth.expiry_date < moment().valueOf() + 60000
+  ) {
+    oauth2Client.credentials.refresh_token =
+      oauth2Client.credentials.refresh_token || values.oauth.refresh_token;
+    const result = await oauth2Client.getAccessToken();
+    await saveTokens(result.res.data);
+    return {
+      ...values,
+      oauth: result.res.data,
+    };
+  }
+  return values;
+};
+
+const storeHistoryIdInDatabase = async (historyId: any) => {
+  return await db.ref(`/gmail/coordinator`).update({
+    lastSyncHistoryId: historyId,
+  });
+};
+
+export const initWatch = functions.https.onRequest(
+  async (req: functions.Request, res: functions.Response) => {
+    // Initiate gmail client
+    const { oauth } = await fetchToken();
+    oauth2Client.setCredentials({
+      access_token: oauth.access_token,
+      refresh_token: oauth.refresh_token,
+    });
+    const gmail = await Google.google.gmail({
+      version: 'v1',
+      auth: oauth2Client,
+    });
+
+    try {
+      // Make sure we have the correct labelId because GMail internally gives a random one that is
+      // not based on the given name, so we need to filter it by the given name
+      const labelResults = await gmail.users.labels.list({
+        userId: 'me',
+      });
+      const doneLabelObj = labelResults.data.labels.filter(label => {
+        return label.name === 'SQRDone';
+      })[0];
+
+      const watchResults = await gmail.users.watch({
+        userId: 'me',
+        requestBody: {
+          labelIds: [doneLabelObj.id],
+          topicName: 'projects/audio-seva-team-test/topics/gmail-labeled-done',
+        },
+      });
+
+      if (watchResults.data && doneLabelObj) {
+        // Store the most recent historyId
+        await storeHistoryIdInDatabase(watchResults.data.historyId);
+        return res
+          .status(200)
+          .send(
+            `Watch (${watchResults.data.historyId}) created on ${
+              doneLabelObj.name
+            }/${doneLabelObj.id} for coordinator`
+          );
+      } else {
+        return res
+          .status(301)
+          .send(`Something went wrong, check logs for Gmail-initWatch`);
+      }
+    } catch (error) {
+      console.error(error);
+      return res
+        .status(301)
+        .send(`Something went wrong, check logs for Gmail-initWatch`);
+    }
+  }
+);
+
+export const doneHandler = functions.pubsub
```

Comment:
Earlier I requested for the following name:
```suggestion
export const processMarkingSubmissionAsDone = functions.pubsub
```

Reply:
Changed.